src/internal/utils.c

```diff
@@ -93,3 +94,96 @@ igraph_error_t igraph_i_matrix_subset_vertices(
 
     return IGRAPH_SUCCESS;
 }
+
+
+/* Lexicographic edge comparator, used with igraph_qsort() in igraph_i_simplify_edge_list() */
+static int edge_comparator(const void *a, const void *b) {
+    igraph_integer_t *A = (igraph_integer_t *) a;
+    igraph_integer_t *B = (igraph_integer_t *) b;
+
+    if (A[0] < B[0]) {
+        return -1;
+    }
+    if (A[0] > B[0]) {
+        return  1;
+    }
+
+    /* first are equal */
+    if (A[1] < B[1]) {
+        return -1;
+    }
+    if (A[1] > B[1]) {
+        return  1;
+    }
+
+    /* second are equal, so the edges must be equal */
+    return 0;
+}
+
+/**
+ * Simplify an edge list in-place. Edges may be reordered by this function.
+ *
+ * TODO: Refactor this to take the number of vertices as input and use linear-time radix sort.
+ *
+ * \param edges The edge list vector, as a consecutive list of pairs. It will be modified in-place.
+ * \param self_loops Set to \c false to remove self-loops.
+ * \param multi_edges Set to \c false to eliminate multi-edges.
+ * \param directed Whether to treat edges as directed.
+ * \return Error code.
+ */
+igraph_error_t igraph_i_simplify_edge_list(
+        igraph_vector_int_t *edges,
+        igraph_bool_t self_loops, igraph_bool_t multi_edges,
+        igraph_bool_t directed) {
+
+    igraph_integer_t size = igraph_vector_int_size(edges);
+
+    if (size == 0) {
+        return IGRAPH_SUCCESS;
+    }
+
+    /* Canonicalize undirected edges. */
+    if (!directed) {
+        for (igraph_integer_t i = 0; i < size; i += 2) {
+            if (VECTOR(*edges)[i] > VECTOR(*edges)[i + 1]) {
+                igraph_integer_t temp = VECTOR(*edges)[i];
+                VECTOR(*edges)[i] = VECTOR(*edges)[i + 1];
+                VECTOR(*edges)[i + 1] = temp;
+            }
+        }
+    }
+
+    /* Sort edge list. Not needed if multi edges are allowed. */
+    if (!multi_edges) {
+        igraph_qsort(VECTOR(*edges), size / 2, 2 * sizeof(igraph_integer_t), &edge_comparator);
+    }
+
+    /* Remove self-loops and duplicate edges */
+
+    igraph_integer_t last_added = 0;
+    igraph_integer_t skipped = 0;
+
+    if (!self_loops) {
+        for (igraph_integer_t i = 0; i < size; i += 2) {
+            if (VECTOR(*edges)[i] == VECTOR(*edges)[i + 1]) {
+                skipped += 1;
+            } else {
+                break;
+            }
+        }
+    }
+
+    for (igraph_integer_t i = skipped * 2 + 2 ; i < size; i += 2) {
```

Comment:
This is pretty convoluted; please add a comment explaining how it works. I _think_ I understand why it is correct now, but I had to stop for several minutes to think and even to try and find a counterexample (which I eventually gave up on).

Renaming `self_loops` to `allow_self_loops` and `multi_edges` to `allow_multi_edges` helps a bit with the interpretation of the big `if` condition, but it still needs an explanation.

Reply:
@Zepeacedust Can you please address this, and use `igraph_bool_t remove_multiple, igraph_bool_t remove_loops` instead, just like in `igraph_simplify()`? This means both renaming the parameters and negating the conditions.